hash.go

```diff
@@ -396,21 +401,7 @@ type sha1Marshal struct {
 }
 
 func (h *sha1Marshal) MarshalBinary() ([]byte, error) {
-	d := (*sha1State)(h.hashState())
-	if d == nil {
-		return nil, errors.New("crypto/sha1: can't retrieve hash state")
-	}
-	b := make([]byte, 0, sha1MarshaledSize)
-	b = append(b, sha1Magic...)
-	b = appendUint32(b, d.h[0])
-	b = appendUint32(b, d.h[1])
-	b = appendUint32(b, d.h[2])
-	b = appendUint32(b, d.h[3])
-	b = appendUint32(b, d.h[4])
-	b = append(b, d.x[:d.nx]...)
-	b = b[:len(b)+len(d.x)-int(d.nx)] // already zero
-	b = appendUint64(b, uint64(d.nl)>>3|uint64(d.nh)<<29)
-	return b, nil
+	return h.AppendBinary(nil)
```

Comment:
We also know the buffer size here.

```suggestion
	buf := make([]byte, 0, sha1MarshaledSize)
	return h.AppendBinary(buf)
```

Reply:
fixed